### PR TITLE
Add performance benchmark for shaping, get extents and draw

### PR DESCRIPTION
**Pull request by ebraminio**
`rm -rf build && meson build -Dbenchmark=true -Dexperimental_api=true -Doptimization=2 && ninja -Cbuild && build/perf/perf`

gives
```
---------------------------------------------------------------------------------------
Benchmark                                                Time           CPU Iterations
---------------------------------------------------------------------------------------
shape/fa-thelittleprince.txt - Amiri              72493871 ns   72339936 ns          9
shape/fa-thelittleprince.txt - NotoNastaliqUrdu  276136994 ns  275646177 ns          2
shape/fa-monologue.txt - Amiri                    12534512 ns   12508287 ns         58
shape/fa-monologue.txt - NotoNastaliqUrdu         49263441 ns   49196080 ns         13
shape/en-thelittleprince.txt - Roboto             34517598 ns   34475158 ns         20
shape/en-words.txt - Roboto                       40218073 ns   40162358 ns         17
extents/cff - ot - SourceSansPro                   2943239 ns    2939081 ns        238
extents/cff - ft - SourceSansPro                   7115167 ns    7105228 ns         97
extents/cff2 - ot - AdobeVFPrototype               2898143 ns    2894036 ns        241
extents/cff2 - ft - AdobeVFPrototype               2188817 ns    2186489 ns        317
extents/cff2/vf - ot - AdobeVFPrototype            3016606 ns    3013072 ns        232
extents/cff2/vf - ft - AdobeVFPrototype            2258966 ns    2256708 ns        309
extents/glyf - ot - SourceSerifVariable              39210 ns      39171 ns      18584
extents/glyf - ft - SourceSerifVariable            1062746 ns    1061467 ns        655
extents/glyf/vf - ot - SourceSerifVariable         4181210 ns    4176616 ns        168
extents/glyf/vf - ft - SourceSerifVariable         3770242 ns    3758127 ns        186
extents/glyf - ot - Comfortaa                        22834 ns      22734 ns      30695
extents/glyf - ft - Comfortaa                       614056 ns     611319 ns       1122
extents/glyf/vf - ot - Comfortaa                   2751386 ns    2738147 ns        259
extents/glyf/vf - ft - Comfortaa                   2348252 ns    2336775 ns        299
extents/glyf - ot - Roboto                           33721 ns      33570 ns      20162
extents/glyf - ft - Roboto                          764178 ns     760421 ns        912
draw/cff - ot - SourceSansPro                      3763109 ns    3744137 ns        188
draw/cff - ft - SourceSansPro                      7156332 ns    7120231 ns         96
draw/cff2 - ot - AdobeVFPrototype                  3042498 ns    3027313 ns        230
draw/cff2 - ft - AdobeVFPrototype                  2223677 ns    2213039 ns        316
draw/cff2/vf - ot - AdobeVFPrototype               3251870 ns    3236131 ns        216
draw/cff2/vf - ft - AdobeVFPrototype               2241302 ns    2230490 ns        314
draw/glyf - ot - SourceSerifVariable               3091201 ns    3076134 ns        228
draw/glyf - ft - SourceSerifVariable               1064737 ns    1059445 ns        655
draw/glyf/vf - ot - SourceSerifVariable            5414722 ns    5387867 ns        127
draw/glyf/vf - ft - SourceSerifVariable            3809064 ns    3788532 ns        174
draw/glyf - ot - Comfortaa                         1991267 ns    1981326 ns        350
draw/glyf - ft - Comfortaa                          618730 ns     615736 ns       1113
draw/glyf/vf - ot - Comfortaa                      3496356 ns    3479246 ns        201
draw/glyf/vf - ft - Comfortaa                      2381006 ns    2367965 ns        298
draw/glyf - ot - Roboto                            1777238 ns    1770355 ns        390
draw/glyf - ft - Roboto                             748675 ns     747645 ns        917
```

(comparing with freetype shouldn't be fair as the different API, linking status and optimization and things but is given as an illustration of difference between different functionalities. hb's `draw`, as provide close path and does some sanitization can be optimized easily)

### CODE REVIEW

**Comment by khaledhosny** [comment]
Hmm, I wonder why is Amiri ~2 times slower than NotoNastaliqUrdu.

**Comment by ebraminio** [comment]
> Hmm, I wonder why is Amiri ~2 times slower than NotoNastaliqUrdu.

Feel free to let me know any other test or combination could be tested here, thanks.

**Comment by ebraminio** [comment]
https://github.com/marketplace/actions/continuous-benchmark

File: perf/perf-shaping.cc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  assert (hb_blob_get_length (blob));
+  hb_face_t *face = hb_face_create (blob, 0);
+  hb_font_t *font = hb_font_create (face);
+  hb_face_destroy (face);
+
+  hb_blob_t *text_blob = hb_blob_create_from_file (text_path);
+  assert (hb_blob_get_length (text_blob));
+  unsigned text_blob_length;
+  const char *text = hb_blob_get_data (blob, &text_blob_length);
+  assert (text_length < text_blob_length);
```
Comment (ebraminio) [inline perf/perf-shaping.cc:20]: I am not shaping the whole text here, maybe I should but that means not iteration and shaping only once.

**Comment by ebraminio** [comment]
> Hmm, I wonder why is Amiri ~2 times slower than NotoNastaliqUrdu.

Apparently something is wrong here, investigating.

**Comment by ebraminio** [comment]
> Hmm, I wonder why is Amiri ~2 times slower than NotoNastaliqUrdu.

I was using a wrong blob as input 🤦 fixed now,

**Comment by khaledhosny** [comment]
> > Hmm, I wonder why is Amiri ~2 times slower than NotoNastaliqUrdu.
> 
> I was using a wrong blob as input 🤦 fixed now.

I see, you had me little worried!

**Comment by behdad** [comment]
First, THANK YOU Ebrahim for working on this.  This has been the most glaring missing piece from our infrastructure.  I'll look into and comment soon.

File: perf/perf-draw.hh [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    FT_Outline_Funcs draw_funcs;
+    draw_funcs.move_to = (FT_Outline_MoveToFunc) _ft_move_to;
+    draw_funcs.line_to = (FT_Outline_LineToFunc) _ft_line_to;
+    draw_funcs.conic_to = (FT_Outline_ConicToFunc) _ft_conic_to;
+    draw_funcs.cubic_to = (FT_Outline_CubicToFunc) _ft_cubic_to;
+    draw_funcs.shift = 0;
+    draw_funcs.delta = 0;
+
+    for (auto _ : state)
```
Comment (ebraminio) [inline perf/perf-draw.hh:71]: just a quick note, actual measuring happens inside this `for (auto _ : state)` and the rest is preparation,

**Comment by behdad** [comment]
So, the first feedback: I want a commandline tool that takes parameters for what to test. It should take font file, text or glyph range or other info, as well as repeat count, fontfunc to use, etc. This should be a commandline tool.

The the test suit should specify the tests in a text file, like we do with our shaping test suite.

That's also similar to what `ftbench` does. Using Google perftool or something else, there was a package with a name like `grieger` but I cannot find (@garretrieger do you remember by any chance?!) that had code for the timing and statistics processing of reporting the timing.  We need that in the commandline tool.

**Comment by ebraminio** [comment]
> So, the first feedback: I want a commandline tool that takes parameters for what to test. It should take font file, text or glyph range or other info, as well as repeat count, fontfunc to use, etc. This should be a commandline tool.
> 
> The the test suit should specify the tests in a text file, like we do with our shaping test suite.
> 
> That's also similar to what `ftbench` does. Using Google perftool or something else, there was a package with a name like `grieger` but I cannot find (@garretrieger do you remember by any chance?!) that had code for the timing and statistics processing of reporting the timing. We need that in the commandline tool.

Is possible, but maybe we can have this first?

**Comment by behdad** [comment]
> Is possible, but maybe we can have this first?

Is you commit to finishing it to a reasonable state instead of just dropping into codebase and move on, yes.

**Comment by ebraminio** [comment]
> Is you commit to finishing it to a reasonable state instead of just dropping into codebase and move on, yes.

I did it the last time I added perf and now here we are, so :)

**Comment by behdad** [comment]
Sounds good to me. Thanks!

**Comment by ebraminio** [comment]
Just to add, have you seen
https://github.com/rhysd/github-action-benchmark
I actually have integrated this but disabled it as it write to gh-pages branch

Our https://github.com/harfbuzz/harfbuzz/blob/master/perf/run.sh even is able to achieve something similar, I will provide the tool and waiting for @garretrieger of course sometime as it is easy to get but let's don't encourage adhoc benchmarking and increase size of our benchmarking suite over the time instead encouraging adhoc ones, just like the project is doing with tests.

But please have a look at the first link first, it provides a similar experience to chromeperf thing we were dreaming about https://chromeperf.appspot.com/group_report?keys=agxzfmNocm9tZXBlcmZyFAsSB0Fub21hbHkYgIDIqI_uwQkM

**Comment by ebraminio** [comment]
#2506

**Pull request merged**